:sparkles: Feat: Add example of Tevm usage in a Next.js app

0xpolarzero opened the pull request:
## Description

Add a Next.js app using Tevm as an example of integration (+ WhatsABI, viem).

See the Readme directly for a comprehensive overview, and the additional information below as well.

Please tell me if there is something wrong. I sometimes lack the vocabulary in English, on EVM-related stuff or even just in computer systems, so there might be some inaccuracies; feel free to modify/correct anything without notice.

See #952.

## Testing

❌

## Additional Information

- [x] I read the [contributing docs](../docs/contributing.md) (if this is your first contribution)

A few notes/limitations on this example app:

1. The app is setup to use prettier + eslint because rome was not enough; i.e. to format and sort Tailwind classes. I actually started with rome before noticing this. Let me know if you would like me to revert to just rome for better consistency with the rest of the repo.

2. Since we want to use Tevm on the browser (e.g. to access local storage), we need to expose API keys to the client; it's ok here, but that might be problematic. It could actually be updated to use server components and separate the logic. For instance, load/dump the client state from/to local storage on the browser-side, then attach it again to the client on the server, after going through the API route.

3. I might be doing something wrong/overkill with types in the `lib/types` folder, as each type is heavily documented using jsdoc. Which seems off because it looks like types are actually defined twice each time. But maybe it's a good practice actually?

---

Your ENS/address: polarzero.eth



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a search functionality allowing users to search for contracts or externally owned accounts.
  - Added components for account and contract interaction, theme switching, and responsive tooltips.
  - Enhanced user experience with a variety of UI elements including buttons, dropdown menus, and tables.

- **Documentation**
  - Updated README with detailed instructions on using the application and its features.

- **Style**
  - Customized Tailwind CSS configuration for a unique look and feel, including dark theme support.

- **Chores**
  - Configured ESLint, Prettier, and VS Code settings for consistent code formatting and linting.
  - Updated `.gitignore`, `.prettierignore`, and environment variable placeholders for better project maintenance.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

roninjin10 commented on examples/next/src/app/global-error.tsx:
```diff
@@ -0,0 +1,43 @@
+  error: Error & { digest?: string };
+  reset: () => void;
+}) => {
+  console.error(error);
```
```suggestion
```

roninjin10 commented on examples/next/src/lib/utils.ts:
```diff
@@ -0,0 +1,87 @@
+};
+
+// Format a string into valid bytes
+const formatBytes = (value: string | number): Hex => {
```
this is confusing name because it's returning `Hex` instead of a `uint8Array` and it's just typecasting
```suggestion
const getHex = (value: string | number): value is Hex => {
```

roninjin10 commented on examples/next/src/lib/utils.ts:
```diff
@@ -0,0 +1,87 @@
+};
+
+// Format a string into a valid address
+const formatAddress = (value: string | number): Address => {
```
```suggestion
export {getAddress} from 'tevm/utils'
```

roninjin10 commented on examples/next/src/app/address/[account]/page.tsx:
```diff
@@ -0,0 +1,72 @@
+  params: { account: string };
+}) {
+  // Remember when the account has been fetched
+  const [fetched, setFetched] = useState(false);
```
Could this just be derived state instead of a useState and effect?
```suggestion
  const fetched = !!params.account
```

0xpolarzero replied:
Actually, this one is a bit confusing, but it's to prevent the `useEffect` to fire `updateAccount` anytime the client/chain changes. Which would happen when the user switches chains/resets a chain. This one would be fired only on page mount.

But yeah, clearly such a hack indicates a bad design. I'll think of a different way,-; which actually now seems as obvious as removing `updateAccount` in the mentioned cases and let this `useEffect` handle it...

roninjin10 commented on examples/next/package.json:
```diff
@@ -0,0 +1,69 @@
+    "sonner": "^1.4.3",
+    "tailwind-merge": "^2.2.1",
+    "tailwindcss-animate": "^1.0.7",
+    "tevm": "1.0.0-next.41",
```
```suggestion
    "tevm": "workspace:^",
```
We want to use the same version of tevm that is in the monorepo instead of installing from npm

roninjin10 commented on examples/next/src/lib/whatsabi.ts:
```diff
@@ -0,0 +1,88 @@
+
+const { EtherscanABILoader, MultiABILoader, SourcifyABILoader } = loaders;
+
+const ETHERSCAN_API_KEY = process.env.ETHERSCAN_API_KEY || '';
```
agreed shouldn't this throw?

roninjin10 commented on examples/next/src/lib/utils.ts:
```diff
@@ -0,0 +1,87 @@
+import { ABI } from '@shazow/whatsabi/lib.types/abi';
+import { clsx, type ClassValue } from 'clsx';
+import { twMerge } from 'tailwind-merge';
+import { isAddress, isHex } from 'viem';
```
```suggestion
import { isAddress, isHex } from 'tevm/utils';
```

roninjin10 commented on examples/next/src/lib/utils.ts:
```diff
@@ -0,0 +1,87 @@
+import { twMerge } from 'tailwind-merge';
+import { isAddress, isHex } from 'viem';
+
+import { Address, Hex } from '@/lib/types/config';
```
```suggestion
import { Address, Hex } from 'tevm/utils';
```

roninjin10 commented on examples/next/src/lib/types/config.ts:
```diff
@@ -0,0 +1,64 @@
+
+import { Chain, Client } from '@/lib/types/providers';
+
+/**
+ * @type {`0x${string}`} Address
+ * @notice A valid Ethereum address (after checksum verification)
+ */
+export type Address = `0x${string}`;
```
```suggestion
```
Can just import Address from tevm

roninjin10 commented on examples/next/src/lib/types/config.ts:
```diff
@@ -0,0 +1,64 @@
+ * @notice A valid Ethereum address (after checksum verification)
+ */
+export type Address = `0x${string}`;
+
+/**
+ * @type {`0x${string}`} Hex
+ * @notice A valid hexadecimal string
+ */
+export type Hex = `0x${string}`;
```
```suggestion
```
Same this is exported from tevm

roninjin10 commented on examples/next/src/lib/types/config.ts:
```diff
@@ -0,0 +1,64 @@
+ */
+export type Hex = `0x${string}`;
+
+/**
+ * @type {Object} Account
+ * @notice Either a contract or an account, with the result of Tevm `Client.getAccount`
+ * @param {Address} address The address of the account
+ * @param {bigint} balance The balance of the account in native tokens
+ * @param {Hex} deployedBytecode The deployed bytecode of the account if it's a contract
+ * @param {bigint} nonce The nonce of the account
+ * @param {boolean} isContract Whether the account is a contract or not
+ * @param {boolean} isEmpty Whether the account is empty or not (can indicate an uninitialized account)
+ * @param {GetAccountError[]} errors Any errors that occurred when retrieving the state of the account
+ */
+export type Account = {
+  address: Address;
+  balance: bigint | string; // string for local storage
+  deployedBytecode: Hex;
+  nonce: bigint | string; // string for local storage
+  isContract: boolean;
+  isEmpty: boolean;
+  errors?: GetAccountError[];
+};
```
```suggestion
```
Can import this from tevm

roninjin10 commented on examples/next/src/lib/types/config.ts:
```diff
@@ -0,0 +1,64 @@
+  errors?: GetAccountError[];
+};
+
+/**
+ * @type {Object} FunctionAbi
+ * @notice Returned from WhatsABI and formatted (keep only function + add unique id)
+ * @dev We made sure to get a name for any function (at the very worst `function-${index}`)
+ * @param {ABIFunction} The base function's ABI from WhatsABI
+ * @param {string} id A unique identifier for the function
+ * @param {string} name Either the name of the function, selector or a unique identifier
+ */
+export type FunctionAbi = ABIFunction & {
+  id: string;
+  // We made sure to get a name for any function (at the very worst `function-${index}`)
+  name: string;
+};
```
`import { FunctionAbi } from 'tevm/utils'`

0xpolarzero replied:
I'm confused about this. Basically `tevm/utils` exposes:
```ts
type AbiFunction = {
  type: 'function';
  constant?: boolean | undefined;
  gas?: number | undefined;
  inputs: readonly AbiParameter[];
  name: string;
  outputs: readonly AbiParameter[];
  payable?: boolean | undefined;
  stateMutability: AbiStateMutability;
};
```

And WhatsABI will return:

```ts
type ABIFunction = {
  type: 'function'; // TODO: constructor, receive, fallback
  selector: string;
  name?: string;
  outputs?: { type: string; length?: number; name: string }[];
  inputs?: { type: string; name: string }[];
  sig?: string;
  sigAlts?: string[];
  payable?: boolean;
  stateMutability?: StateMutability;
};
```

... because we're not guaranteed to receive any `inputs`, `outputs` and `stateMutability`. And I need an `id` to be sure each function is unique.

Do you mean I should still use the type from Tevm and make sure the retrieved ABI items fit in? Maybe with additional markers to disclose parameters that could not be retrieved.

roninjin10 commented on examples/next/src/lib/tevm.ts:
```diff
@@ -0,0 +1,348 @@
+import { Chain, Client } from '@/lib/types/providers';
+import { ExpectedType, TxResponse } from '@/lib/types/tx';
+
+const alchemyApiKey = process.env.ALCHEMY_API_KEY || '';
```
```suggestion
const defaultApiKey = "_gg7wSSi0KMBsdKnGVfHDueq6xMB9EkC"

const alchemyApiKey = process.env.ALCHEMY_API_KEY || defaultApiKey;
```

roninjin10 commented on examples/next/src/lib/constants/providers.ts:
```diff
@@ -0,0 +1,82 @@
+      rpcUrl: 'https://eth-sepolia.g.alchemy.com/v2/',
+    },
+  },
+  hardhat: {
+    ...foundry,
+    custom: {
+      rpcUrl: 'http://localhost:8545',
+    },
+  },
```
can we add zora too?

roninjin10 reviewed (changes requested):
Thanks for pr did a really quick skim of the code will give it a more thourough reveiew this weekend

0xpolarzero commented:
https://github.com/evmts/tevm-monorepo/pull/991#discussion_r1514493428

> [on the Etherscan API key] agreed shouldn't this throw?

Answering here because I can't comment.

Actually, I was using the same key for all chains, which was wrong. I just fixed this and added correct ones for each Etherscan-compatible explorer API.

But I don't think it should thow; if there is no key, it will use an empty string, and still be able to make a request every 5 seconds per IP (afaik for Etherscan forks). Which seems fine for this use case.

roninjin10 reviewed (changes requested):
Did another quick read of the code. Looks like a great start

0xpolarzero commented on examples/next/src/styles/globals.css:
```diff
@@ -0,0 +1,73 @@
+  }
+}
+
+/* sonner */
+/* TODO this will break the closing animation but should be fixed */
+/* .toaster.group li[data-removed='true'] {
+      display: none;
+} */
```
Absolutely not! Forgot to clean up this one

0xpolarzero commented on examples/next/src/lib/types/providers.ts:
```diff
@@ -0,0 +1,22 @@
+  };
+};
+
+/**
+ * @type {Client}
+ * @notice A Tevm memory client
+ * @dev This will be used to write/read to/from the chain's state.
+ */
+export type Client = MemoryClient;
```
Not really, I just like to keep such types in one file, so it can be updated in one place if we want to change to another client later, instead of skimming through all files.

But maybe we do want to skim through all the files actually. I'll update!

roninjin10 replied:
don't bother was just confused

0xpolarzero commented on examples/next/next.config.mjs:
```diff
@@ -0,0 +1,27 @@
+/** @type {import('next').NextConfig} */
+const nextConfig = {
```
Is this fine, or too convoluted?

```js
// next.config.mjs

// Api keys
const {
  ALCHEMY_API_KEY,
  ETHERSCAN_API_KEY,
  ARBISCAN_API_KEY,
  BASESCAN_API_KEY,
  OPTIMISTIC_ETHERSCAN_API_KEY,
  POLYGONSCAN_API_KEY,
  ZORA_SUPERSCAN_API_KEY,
} = process.env;

const env = {
  ALCHEMY_API_KEY,
  ETHERSCAN_API_KEY,
  ARBISCAN_API_KEY,
  BASESCAN_API_KEY,
  OPTIMISTIC_ETHERSCAN_API_KEY,
  POLYGONSCAN_API_KEY,
  ZORA_SUPERSCAN_API_KEY,
};

Object.keys(env).forEach((key) => {
  if (!env[key]) {
    console.warn(
      `No ${key} environment variable set. You may face throttling.`,
    );
  }
});

/** @type {import('next').NextConfig} */
const nextConfig = {
  ...
  // These will get exposed to the browser
  // If you would like to keep this key private, use an API route/server components to
  // access it and return the data instead
  env,
};

export default nextConfig;
```

roninjin10 replied:
It's fine as long as destructuring process.env like this doesn't cause issues. for most frameworks you need to reference process.env.FOO directly because webpack or vite replace the environment variables in place (process.env doesn't exist in the browser)

0xpolarzero commented:
https://github.com/evmts/tevm-monorepo/pull/991#discussion_r1515647731
> Oh I see so this is the whatsabi type instead of the tevm/viem/abitype type. Does whatsabi export this type? It would be more clear if we imported it from whatsabi

Ok! Right here I'm just adding a unique `id` (to be able to refocus inputs with a unique accessor) and a required `name` (it's already in `ABIFunction` but it's optional); here we want to make sure there is always a name to display.

### On the name

See this:

```ts
// lib/utils.ts

export const getFunctionName = (funcOrEvent: ABI[number], index: number) => {
  return funcOrEvent.type === 'event'
    ? funcOrEvent.name
    : funcOrEvent.name ||
        funcOrEvent.sig ||
        funcOrEvent.selector ||
        `function-${index.toString()}`;
};
```

### On the id

Here it's a bit messed up... Basically [right here] we have a table that displays the abi.

Anytime we type in an input, it calculates if the call button can be clicked... which is part of the table, so it will rerender. So we want to remember the currently focused input, to be able to refocus it on rerender. Hence the unique id—I feel like relying on the name or rest of the function might cause issue, e.g. if the contract has overloading functions.

Will think about it

0xpolarzero commented:
https://github.com/evmts/tevm-monorepo/pull/991#discussion_r1515641284
> Can we get rid of the useEffect? I generally am a nit for never using useEffects.

Agree with that, I'll see the ones I can get rid of. I did refactor this one, but I believe we need to keep it in case the user either directly lands on this page or reloads? Because we need to pass the query params (the address) to the function.

0xpolarzero commented:
I got rid of the useEffects. I replaced them with useMemo when it would make sense, I don't know if it's better or kind of the same under the hood?

Basically, there are only 3 left:
- one [here in use-media-query](https://github.com/0xpolarzero/tevm-monorepo/blob/example-next/examples/next/src/lib/hooks/use-media-query.tsx#L12) and another one [here in theme-switcher](https://github.com/0xpolarzero/tevm-monorepo/blob/example-next/examples/next/src/components/config/theme-switcher.tsx#L16), because these make sense to me (listening to events, hooks)
- another one [here in elapsed-time](https://github.com/0xpolarzero/tevm-monorepo/blob/example-next/examples/next/src/components/common/elapsed-time.tsx#L31) because it starts an interval we want to get rid of on unmount.

Let me know if the useMemo is not ideal as well, and I should think about other ways.

0xpolarzero commented:
I had to add back the `useEffect` on the account page, even though I really wanted to get rid of it. But I could not find another solution to react to page change only on navigation (previous/back), and not on other actions (search, reload), which are already handled.

Basically, the code is as follows:
```ts
useEffect(() => {
  if (
    client &&
    account &&
    isAddress(params.account) &&
    account.address !== params.account
  )
    updateAccount(params.account as Address, {
      updateAbi: true,
      chain,
      client,
    });
}, [params.account, account, chain, client, updateAccount]);
```

Basically, this will trigger the update strictly only on page change (/address/0x1 => /address/0xB). The rest is taken care of individually (first mount is handled during `useProviderStore` hydration, changing/resetting chains in their own component.

The search function will now trigger an update only in the specific case where we're clicking the example button on the page of the example account, on a different chain (it wants to update only if the chain changes).

To try to make it clearer:
1. Launch the app, reload a page -> account is fetched during hydration, along with the provider ([see here](https://github.com/0xpolarzero/tevm-monorepo/blob/example-next/examples/next/src/lib/store/use-provider.ts#L167)).
2. Change chains, reset a chain -> account is fetched along with the new provider ([reset](https://github.com/0xpolarzero/tevm-monorepo/blob/example-next/examples/next/src/components/core/selection/chain.tsx#L81), [change](https://github.com/0xpolarzero/tevm-monorepo/blob/example-next/examples/next/src/components/core/selection/chain.tsx#L123)).
3. Enter an address and click search, navigate back/forward with the browser -> account is fetched on mount ([here](https://github.com/0xpolarzero/tevm-monorepo/blob/example-next/examples/next/src/app/address/%5Baccount%5D/page.tsx#L47)).

roninjin10 commented:
Should be able to get to this this weekend? Is it ok if I make large edits to this?

0xpolarzero commented:
@roninjin10 Of course, please edit anything you want! Honestly, that's a great feedback opportunity for me.

Don't hesitate if you feel like just deleting some bunch of code that feels off/unnecessary as well. I won't be offended.

roninjin10 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/991?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#991** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/991?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#1010** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1010?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1009** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1009?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1008** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1008?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1007** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1007?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1006** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1006?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1005** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1005?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1004** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1004?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1003** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1003?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1002** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1002?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1001** <a href="https://app.graphite.dev/github/pr/evmts/tevm-monorepo/1001?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @0xpolarzero and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->
